diesel/src/sqlite/connection/stmt.rs

```diff
@@ -126,18 +220,159 @@ impl Drop for Statement {
     }
 }
 
+// A warning for future editiors:
+// Changing this code to something "simplier" may
+// introduce undefined behaviour. Make sure you read
+// the following discussions for details about
+// the current version:
+//
+// * https://github.com/weiznich/diesel/pull/7
+// * https://users.rust-lang.org/t/code-review-for-unsafe-code-in-diesel/66798/
+// * https://github.com/rust-lang/unsafe-code-guidelines/issues/194
+struct BoundStatement<'stmt, 'query> {
+    statement: MaybeCached<'stmt, Statement>,
+    // we need to store the query here to ensure noone does
+    // drop it till the end ot the statement
+    // We use a boxed queryfragment here just to erase the
+    // generic type, we use NonNull to communicate
+    // that this is a shared buffer
+    query: Option<NonNull<dyn QueryFragment<Sqlite> + 'query>>,
+    // we need to store any owned bind values speratly, as they are not
+    // contained in the query itself. We use NonNull to
+    // communicate that this is a shared buffer
+    binds_to_free: Vec<(i32, Option<NonNull<[u8]>>)>,
+}
+
+impl<'stmt, 'query> BoundStatement<'stmt, 'query> {
+    fn bind<T>(
+        statement: MaybeCached<'stmt, Statement>,
+        query: T,
+    ) -> QueryResult<BoundStatement<'stmt, 'query>>
+    where
+        T: QueryFragment<Sqlite> + QueryId + 'query,
+    {
+        // Don't use a trait object here to prevent using a virtual function call
+        // For sqlite this can introduce a measurable overhead
+        let query = Box::new(query);
```

Comment:
```suggestion
        // Query is boxed here to make sure it won't move in memory anymore, so any bind
        // it could output would stay valid.
        let query = Box::new(query);
```